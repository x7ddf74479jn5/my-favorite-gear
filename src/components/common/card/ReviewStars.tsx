import StarIcon from "@mui/icons-material/Star";
import StarHalfIcon from "@mui/icons-material/StarHalf";
import type { FC } from "react";
import React, { memo } from "react";

import type { Gear } from "@/services/models/gear";

type ReviewStarsProps = Pick<Gear, "reviewAverage">;

const ReviewStars: FC<ReviewStarsProps> = ({ reviewAverage }) => {
  if (!reviewAverage || reviewAverage > 5) return null;

  const integerPart = Math.floor(reviewAverage);
  const decimalPart = reviewAverage - integerPart;
  let starCount =
    decimalPart < 0.25
      ? integerPart
      : decimalPart < 0.5
      ? integerPart + decimalPart
      : decimalPart < 0.75
      ? integerPart + decimalPart
      : Math.ceil(reviewAverage);
  const stars = [];
  while (starCount > 0) {
    if (starCount < 1) {
      stars.push(<StarHalfIcon key={starCount} />);
      break;
    }
    stars.push(<StarIcon key={starCount} />);
    starCount--;
  }

  return <>{stars}</>;
};

export default memo(ReviewStars);
