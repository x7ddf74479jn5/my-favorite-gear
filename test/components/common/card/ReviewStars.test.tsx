import React from "react";
import { describe, expect, it } from "vitest";

import ReviewStars from "@/components/common/card/ReviewStars";

import { render } from "../../../test-utils";

describe("ReviewStars", () => {
  it("should match snapshot", () => {
    const renderResult = render(<ReviewStars reviewAverage={1} />);
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it("should render correctly", () => {
    const renderResult = render(<ReviewStars reviewAverage={1} />);
    expect(renderResult.container).not.toBeEmptyDOMElement();

    renderResult.rerender(<ReviewStars reviewAverage={1.5} />);
    expect(renderResult.container).not.toBeEmptyDOMElement();
  });

  it("should not render if provided invalid arguments", () => {
    const renderResult = render(<ReviewStars reviewAverage={0} />);
    expect(renderResult.container).toBeEmptyDOMElement();

    renderResult.rerender(<ReviewStars reviewAverage={5.5} />);
    expect(renderResult.container).toBeEmptyDOMElement();
  });
});
