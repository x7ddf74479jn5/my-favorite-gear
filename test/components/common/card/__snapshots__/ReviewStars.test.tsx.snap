// Vitest Snapshot v1

exports[`ReviewStars > should match snapshot 1`] = `
<DocumentFragment>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
    data-testid="StarIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
    />
  </svg>
</DocumentFragment>
`;

exports[`ReviewStars should match snapshot 1`] = `
<DocumentFragment>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
    />
  </svg>
</DocumentFragment>
`;
